### PR TITLE
Transport refactor

### PR DESCRIPTION
**Pull request by boks1971**
Attempt to clean up the transport related stuff from `ParticipantImpl`
- A new module `TransportManager` which manages the peer connections.
- All peer connection related stuff moved into `PCTransport`
  - No more direct access of peer connection object (code like `p.subscriber.pc` or `p.publisher.pc` no more)
  - Data channel objects maintained inside `PCTransport`
  - All candidate filtering happens internally in `PCTransport`
 - `TransportManager` provides APIs which are labeled as `Subscriber`, `Publisher`, `PrimaryTransport`, `AnyTransport` to make it clear what is being requested/operated on.

### CODE REVIEW

File: pkg/rtc/participant_internal_test.go [hunk 1]
```diff
@@ -678,17 +680,28 @@ func TestDisableCodecs(t *testing.T) {
...
 	err = participant.HandleOffer(sdp)
 	require.NoError(t, err)
+
+	testutils.WithTimeout(t, func() string {
+		if answerReceived.Load() {
+			return ""
+		} else {
+			return "answer not received"
+		}
+	})
```
Comment (boks1971) [inline pkg/rtc/participant_internal_test.go:702]: Introduced this as the `answer` is sent in a go routine now.

File: pkg/rtc/participant.go [hunk 2]
```diff
@@ -1355,112 +1067,55 @@ func (p *ParticipantImpl) handleDataMessage(kind livekit.DataPacket_Kind, data [
...
+		p.disconnectTimer = nil
+		p.lock.Unlock()
+
+		if p.isClosed.Load() || p.State() == livekit.ParticipantInfo_DISCONNECTED {
+			return
+		}
+		p.params.Logger.Infow("closing disconnected participant")
+		_ = p.Close(true, types.ParticipantCloseReasonPeerConnectionDisconnected)
+	})
+	p.lock.Unlock()
```
Comment (boks1971) [inline pkg/rtc/participant.go:1118]: This timer was set up to run only on primary transport. I changed it to run on any transport as we are closing signal connection on any transport failure. So, it made to run this on any transport failure. Also, if any transport fails, that would be bad. So, running this on any transport failure.

File: pkg/rtc/transportmanager.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+		} else if dcExisting {
+			t.params.Logger.Debugw("existing data channel during migration", "label", dcLabel, "id", dcID)
+		} else {
+			t.params.Logger.Debugw("create migrated data channel", "label", dcLabel, "id", dcID)
+		}
+	}
+}
+
+/*
+func (p *ParticipantImpl) incActiveCounter() {
```
Comment (cnderrauber) [inline pkg/rtc/transportmanager.go:496]: can remove these

**Review by cnderrauber — APPROVED** [review]

File: pkg/rtc/transportmanager.go (hunk 3 above)
Comment (boks1971) [reply pkg/rtc/transportmanager.go:496]: Yep, sorry missed those. Thank you for catching. Will remove.

File: pkg/rtc/transport.go [hunk 4]
```diff
@@ -327,20 +476,146 @@ func (t *PCTransport) AddICECandidate(candidate webrtc.ICECandidateInit) error {
...
-	t.params.Logger.Infow("add candidate ", "candidate", candidate.Candidate)
+	t.lock.RLock()
+	if t.preferTCP && !strings.Contains(candidate.Candidate, "tcp") {
+		t.lock.RUnlock()
+		t.params.Logger.Infow("filtering out remote candidate", "candidate", candidate.Candidate)
+		return nil
+	}
+	t.lock.RUnlock()
 
+	t.params.Logger.Infow("add candidate ", "candidate", candidate.Candidate)
```
Comment (davidzhao) [inline pkg/rtc/transport.go:487]: unrelated, but to make logs more readable, it'd be great to print all ICE candidates in the same log entry as an array. would be great to have both local and remote candidates in the same entry too.

**Review by davidzhao — APPROVED** [review]
LGTM!

File: pkg/rtc/transport.go (hunk 4 above)
Comment (boks1971) [reply pkg/rtc/transport.go:487]: Okay, will see how to do that. Will look at it when I do caching candidates.

**Pull request merged**